### PR TITLE
Implement multi-port endpoints

### LINKED ISSUE #5895: Check protocol of port defined in service and pod
I created a service with following template

	id: webserver
	kind: Service
	apiVersion: v1beta1
	protocol: UDP
	port: 8888
	containerPort: testport

But in the in pod behind the service the port is defined using TCP protocol 

 	ports:
      - name: testport
        protocol: TCP
        containerPort: 80

I think it should throw an exception rather than allow the service to be created. Though currently the proxy support only TCP protocol.

### PR DESCRIPTION
**Pull request by thockin**
Another piece of multi-port services.

Instead of endpoints being a flat list, it is now a list of "subsets"
where each is a struct of {Addresses, Ports}.  To generate the list of
endpoints you need to take union of the Cartesian products of the
subsets.  This is compact in the vast majority of cases, yest still
represents named ports and corner cases (e.g. each pod has a different
port number).

This also stores subsets in a deterministic order (sorted by hash) to
avoid spurious updates and comparison problems.

Also closes #5895

### CODE REVIEW

**Comment by thockin** [comment]
This is sort of useless on its own.  Once this is in I will finish multi-port Services on top of it.  I am wary to rebase onto this without some basic review.

Caveats: I got this to pass 'make test' and now I am putting it here, with almost no testing.  I'd like to get review started ASAP, as I expect comments and questions.

In the mean time, there are a few things not done yet, whcih i will wrap up into new commits.  I will then do manual testing and run e2e (neither of which I have done yet).

@satnam6502 @quinton-hoole @bgrant0607

**Comment by thockin** [comment]
I have done a bunch of manual testing of this, and it all seems above-board.  There are a few things that I need to wrap up to make it whole, but fundamentally this piece seems to work.

Once we have the other part (multi-port services) it will be possible to write proper e2e tests.  As it is I just did a bunch manually by racing with the system.

I will focus tomorrow on the last few pieces of this, and then rebase the multi-port services on top of this and focus on wrapping that up.

Meantime, please review!

**Comment by thockin** [comment]
Open questions:

* If you have a multi-port service with all ports named, which port do we use
  for /proxy/services/foo?  The packing/sorting process loses the "first-ness" 
  property.  Could carry a flag "isDefault" on endpoint-ports (set from the first service port), or we could
  demand that you say /proxy/services/foo:portname

* We allow /proxy/services/foo:portname but not /proxy/services/foo:portnum - is
  this important?

**Comment by thockin** [comment]
e2e is running overnight.  Going home.

**Comment by bgrant0607** [comment]
Thanks much.

In the case of multiple ports, proxy should require the port to be specified.

File: pkg/api/types.go [hunk 1]
```diff
@@ -916,29 +916,62 @@ type Service struct {
...
+//     Addresses: [{"ip": "10.10.1.1"}, {"ip": "10.10.2.2"}],
+//     Ports:     [{"name": "a", "port": 8675}, {"name": "b", "port": 309}]
+//   }
+// The resulting set of endpoints can be viewed as:
+//     a: [ 10.10.1.1:8675, 10.10.2.2:8675 ],
+//     b: [ 10.10.1.1:309, 10.10.2.2:309 ]
+type EndpointSubset struct {
+	Addresses []EndpointAddress
+	Ports     []EndpointPort
+}
```
Comment (satnam6502) [inline pkg/api/types.go]: Personally I find it confusion to have this called `EndpointSubset` since it is not directly a set and even if it were I am not sure if thinking of it being a subset of the "union of Castesian products set" is that helpful? This type is really a tuple if addresses and ports so why not name it to reflect what it holds e.g. `AddressesAnhdPorts` or `EndpointsSpec` ?

File: pkg/api/types.go [hunk 2]
```diff
@@ -916,29 +916,62 @@ type Service struct {
...
 type Endpoints struct {
 	TypeMeta   `json:",inline"`
 	ObjectMeta `json:"metadata,omitempty"`
 
-	// Optional: The IP protocol for these endpoints. Supports "TCP" and
-	// "UDP".  Defaults to "TCP".
-	Protocol  Protocol   `json:"protocol,omitempty"`
-	Endpoints []Endpoint `json:"endpoints,omitempty"`
+	// The set of all endpoints is the union of all subsets.
+	Subsets []EndpointSubset
```
Comment (satnam6502) [inline pkg/api/types.go]: Since `EndpointSubset` is not directly a set (of course it can be represented as a set) I find the `Set` naming a little confusion (at first glance I thought "has the endpoint specificaiton been unraveled by computation into a flat list or `Set`?"). This filed represents a list of endpoints specifications. So if you renamed `EndpointSubset` to`EndpointsSpec` then this is perhaps:
```
  Specs []EndpointsSpec
```

File: pkg/api/types.go [hunk 3]
```diff
@@ -916,29 +916,62 @@ type Service struct {
...
+//   Subsets: [
+//     {
+//       Addresses: [{"ip": "10.10.1.1"}, {"ip": "10.10.2.2"}],
+//       Ports: [{"name": "a", "port": 8675}, {"name": "b", "port": 309}]
+//     },
+//     {
+//       Addresses: [{"ip": "10.10.3.3"}],
+//       Ports: [{"name": "a", "port": 93}, {"name": "b", "port": 76}]
+//     },
+//  ]
```
Comment (satnam6502) [inline pkg/api/types.go]: And I assume there is no intersection between the Cartesian products of each `EndpointSubset` value otherwise it is an error? Specifically, an IP address must not occur in more than one `EndpointSubset`.
Is it fine to have instead of this:

```
Subsets: [
  {
     Addresses: [{"ip": "10.10.1.1"}, {"ip": "10.10.2.2"}],
     Ports: [{"name": "a", "port": 8675}, {"name": "b", "port": 309}]
    },
]
```

to have this:

```
Subsets: [
  {
     Addresses: [{"ip": "10.10.1.1"}],
     Ports: [{"name": "a", "port": 8675}, {"name": "b", "port": 309}]
    },
    {
     Addresses: [{"ip": "10.10.2.2"}],
     Ports: [{"name": "a", "port": 8675}, {"name": "b", "port": 309}]
    },
]
```

since some algebra tells us that these should be equivalent?

File: pkg/api/types.go (hunk 1 above)
Comment (satnam6502) [inline pkg/api/types.go]: It is an error to have some `Addresses` not no `Ports`?
Or some `Ports` and no `Addresses`?
Is this checked/validated? Perhaps I will find out later but that does not seem to be the case in v1beta1/conversion.gp

File: pkg/api/v1beta1/conversion.go [hunk 4]
```diff
@@ -1181,21 +1181,42 @@ func init() {
...
+				out.Protocol = ProtocolTCP
+			}
+			for i := range in.Subsets {
+				ss := &in.Subsets[i]
+				for j := range ss.Ports {
+					ssp := &ss.Ports[j]
+					if ssp.Name != firstPortName {
+						continue
+					}
+					for k := range ss.Addresses {
```
Comment (satnam6502) [inline pkg/api/v1beta1/conversion.go]: Minor, but I think you can capture `ssa` in the `for` assignment and then use `&ssa` in the body or indeed just still use `ssa.Targetref` because I assume you only do `ssp := &ss.Ports[j]` to avoid copying.

File: pkg/api/v1beta1/defaults.go [hunk 5]
```diff
@@ -93,7 +96,43 @@ func init() {
...
+					}
+					port, err := strconv.Atoi(portStr)
+					if err != nil {
+						glog.Errorf("failed to Atoi(%q)", portStr)
+					}
+					ss := EndpointSubset{
+						Addresses: []EndpointAddress{{IP: host, TargetRef: tgtRef}},
+						Ports:     []EndpointPort{{Protocol: obj.Protocol, Port: port}},
+					}
+					obj.Subsets = append(obj.Subsets, ss)
```
Comment (satnam6502) [inline pkg/api/v1beta1/defaults.go]: Very minor, but I would be tempted to inline `ss` and have one less variable in the world -- even if the variable has my initials.

File: pkg/api/v1beta1/defaults_test.go [hunk 6]
```diff
@@ -61,14 +61,56 @@ func TestSetDefaultSecret(t *testing.T) {
 	}
 }
 
+func TestSetDefaulEndpointsLegacy(t *testing.T) {
```
Comment (satnam6502) [inline pkg/api/v1beta1/defaults_test.go]: Now about a comment saying what this is test?

File: pkg/api/v1beta2/defaults.go [hunk 7]
```diff
@@ -95,7 +97,43 @@ func init() {
 		},
 		func(obj *Endpoints) {
 			if obj.Protocol == "" {
-				obj.Protocol = "TCP"
+				obj.Protocol = ProtocolTCP
```
Comment (satnam6502) [inline pkg/api/v1beta2/defaults.go]: Can any of this code be factored out between v1beta{1,2,3}? There seems to be quite a bit of duplication?

File: pkg/conversion/deep_equal_test.go [hunk 8]
```diff
@@ -66,3 +76,60 @@ func TestEqualities(t *testing.T) {
 		}
 	}
 }
+
+func TestDerivates(t *testing.T) {
```
Comment (satnam6502) [inline pkg/conversion/deep_equal_test.go]: `TestDerivatives` ?

File: pkg/kubectl/resource_printer.go [hunk 9]
```diff
@@ -277,16 +275,35 @@ func (h *HumanReadablePrinter) printHeader(columnNames []string, w io.Writer) er
...
+					}
+					addr := &ss.Addresses[i]
+					list = append(list, fmt.Sprintf("%s:%d", addr.IP, port.Port))
+				}
+			}
+		}
+	}
+	ret := strings.Join(list, ",")
+	if more {
+		ret += "..."
```
Comment (satnam6502) [inline pkg/kubectl/resource_printer.go]: So after three values nothing else is printed?

File: pkg/kubectl/resource_printer_test.go [hunk 10]
```diff
@@ -452,7 +452,11 @@ func TestPrinters(t *testing.T) {
 		"pod":             &api.Pod{ObjectMeta: om("pod")},
 		"emptyPodList":    &api.PodList{},
 		"nonEmptyPodList": &api.PodList{Items: []api.Pod{{}}},
-		"endpoints":       &api.Endpoints{Endpoints: []api.Endpoint{{IP: "127.0.0.1"}, {IP: "localhost", Port: 8080}}},
+		"endpoints": &api.Endpoints{
+			Subsets: []api.EndpointSubset{{
+				Addresses: []api.EndpointAddress{{IP: "127.0.0.1"}, {IP: "localhost"}},
+				Ports:     []api.EndpointPort{{Port: 8080}},
+			}}},
```
Comment (satnam6502) [inline pkg/kubectl/resource_printer_test.go]: It can wait until a later PR but it would be goo to rest the "more" case.

File: pkg/proxy/roundrobin.go [hunk 11]
```diff
@@ -218,33 +238,52 @@ func (lb *LoadBalancerRR) updateAffinityMap(service types.NamespacedName, newEnd
...
+		// We need to build a map of portname -> all ip:ports for that
+		// portname.  Explode Endpoints.Subsets[*] into this structure.
+		portsToEndpoints := map[string][]hostPortPair{}
+		for i := range svcEndpoints.Subsets {
+			ss := &svcEndpoints.Subsets[i]
+			for i := range ss.Ports {
+				port := &ss.Ports[i]
+				for i := range ss.Addresses {
+					addr := &ss.Addresses[i]
+					portsToEndpoints[port.Name] = append(portsToEndpoints[port.Name], hostPortPair{addr.IP, port.Port})
```
Comment (satnam6502) [inline pkg/proxy/roundrobin.go]: Or make a `[]hostPortPair`, append to that, and when done update the `portsToEndpoints` map to avoid repeated map lookups? This should work assuming that port names are unique across "subsets".

File: pkg/proxy/roundrobin_test.go [hunk 12]
```diff
@@ -77,101 +77,208 @@ func TestLoadBalanceFailsWithNoEndpoints(t *testing.T) {
...
+			if haystack[i] == needle {
+				found = true
+				break
+			}
+		}
+		if found == false {
+			return false
+		}
+	}
+	return true
```
Comment (satnam6502) [inline pkg/proxy/roundrobin_test.go]: Needs are small and haystacks are large and this is O(M*N). Fine for small values for `haystack` but if we anticipate a larger list then perhaps `sort.Search` might be better?

File: pkg/registry/service/rest.go [hunk 13]
```diff
@@ -233,19 +234,43 @@ var _ = rest.Redirector(&REST{})
...
+	parts := strings.Split(id, ":")
+	if len(parts) > 2 {
+		return nil, nil, errors.NewBadRequest(fmt.Sprintf("invalid service request %q", id))
+	}
+	svcName := parts[0]
+	portStr := ""
+	if len(parts) == 2 {
+		portStr = parts[1]
+	}
+
```
Comment (satnam6502) [inline pkg/registry/service/rest.go]: What if `len(parts) == 1`? Does the loop below make sense for `portStr = ""` ?

Comment (satnam6502) [inline pkg/registry/service/rest.go]: So if a port is not specified do we need to do anything here for backward compatibility? i.e. assume (and check) just one available port and use that?

File: pkg/api/types.go (hunk 3 above)
Comment (satnam6502) [inline pkg/api/types.go]: Can the name of a port be a string which represents number e.g. `"6502"`?
That would be confusing?

Comment (satnam6502) [inline pkg/api/types.go]: Having read more code I see that the repacker will take care of this.

File: pkg/api/types.go (hunk 1 above)
Comment (bgrant0607) [inline pkg/api/types.go]: Currently it is an error to have no ports.

It should be an error to have no addresses.

Re. use of the term "set" -- fair enough. 

Possible alternatives:
* EndpointPortGroup -- it's a group identified by the ports
* EndpointSublist -- less explanatory, but also less likely to be misinterpreted

Comment (bgrant0607) [inline pkg/api/types.go]: BTW, Spec doesn't work because this isn't really desired state, and we also don't use "Spec" to qualify subobjects other than the top-level object Spec subobjects (PodSpec, ReplicationControllerSpec, etc.).

File: pkg/api/types.go (hunk 3 above)
Comment (bgrant0607) [inline pkg/api/types.go]: The intent is that Subset is effectively a map (represented as a list, as everywhere else in the API), where the key is the concatenation of EndpointPort objects.

Comment (satnam6502) [inline pkg/api/types.go]: Ah, and there is even a test for this case in pkg/api/endpoints/util_test.go

File: pkg/registry/service/rest.go (hunk 13 above)
Comment (satnam6502) [inline pkg/registry/service/rest.go]: Tim explained to me what happens here so now I see what it means for `portStr` to be empty (plus associated question of what we should really do in this case).

File: pkg/api/types.go (hunk 1 above)
Comment (thockin) [inline pkg/api/types.go]: On validation, you're right that this should probably be validated.  As it stands, if either addresses or ports is empty, you get an empty result and a success code.  I'll fix it.

On naming: It is a compact representation of a set of tuples, which is but a piece of a larger set == subset.  I'm fine renaming it as long as the rename is more valuable.

Sublist is no more helpful than Subset IMO.

EndpointGroup?  EndpointPortGroup is no more descriptive that EndpointGroup to me.  Keep in mind that this DOES show up in the API that the user experiences in the form of the field name.

I'm open to better answers.  You guys fight it out.

Comment (thockin) [inline pkg/api/types.go]: EndpointMap (field name Maps) ?

File: pkg/api/types.go (hunk 3 above)
Comment (thockin) [inline pkg/api/types.go]: This made me test something I missed and find a bug.  Thanks.

File: pkg/api/v1beta1/defaults.go (hunk 5 above)
Comment (thockin) [inline pkg/api/v1beta1/defaults.go]: ok

File: pkg/api/v1beta1/defaults_test.go (hunk 6 above)
Comment (thockin) [inline pkg/api/v1beta1/defaults_test.go]: k

File: pkg/api/v1beta2/defaults.go (hunk 7 above)
Comment (thockin) [inline pkg/api/v1beta2/defaults.go]: in general we want each API version to be siloed so they can evolve independently.

File: pkg/conversion/deep_equal_test.go (hunk 8 above)
Comment (thockin) [inline pkg/conversion/deep_equal_test.go]: From a different PR, fixed there, if I recall

File: pkg/kubectl/resource_printer.go (hunk 9 above)
Comment (thockin) [inline pkg/kubectl/resource_printer.go]: Correct.  I consulted with Sam and he thought that was a good compromise, rather than printing an arbitrary list.  I could change 3 to 5 if people care.

File: pkg/proxy/roundrobin_test.go (hunk 12 above)
Comment (thockin) [inline pkg/proxy/roundrobin_test.go]: This is a test - don't care (but N and M are small)

File: pkg/registry/service/rest.go (hunk 13 above)
Comment (thockin) [inline pkg/registry/service/rest.go]: I'm fine with requiring port names in the case that all service ports are named.  If users scream we can add an "is default" flag.

File: pkg/api/types.go (hunk 1 above)
Comment (satnam6502) [inline pkg/api/types.go]: Regarding a better name for `Subset` -- I don't have a very strong view one this -- although this is a public field (as you point out). As for `EndpointSublist` -- it seems odd to call something that is a pair a list so this could be confusing.

Comment (bgrant0607) [inline pkg/api/types.go]: Map is even more confusing given that it's not actually represented as a map.

Which is also the argument against set, except that we do de-duplication and it isn't order-preserving, so set seems reasonably appropriate, and there is no such thing as a json set, so no chance for confusion there.

**Comment by thockin** [comment]
Feedback addressed in a new commit, PTAL, especially at the repacking logic and the tests thereof.  If there is any test you can think of that it is missing, holla.

**Comment by thockin** [comment]
e2e passes

**Comment by satnam6502** [comment]
I've looked over the parts that I think changed (with a focus on `RepackSubsets`) and I am happy to declare LGTM (and not quibble about `Subsets`). I am sure we can add more tests later although I can't think of any obvious one off the top of my head (apart from the ones already mentioned).
@bgrant0607 or @quinton-hoole : would either of you like to offer more feedback? If not I will apply the LGTM label.

**Comment by bgrant0607** [comment]
I'll take another look tonight, but am happy to merge and iterate.

File: pkg/api/endpoints/util.go [hunk 14]
```diff
@@ -0,0 +1,154 @@
...
+func RepackSubsets(subsets []api.EndpointSubset) []api.EndpointSubset {
+	// First map each unique port definition to the sets of hosts that
+	// offer it.  The sets of hosts must be de-duped, using IP as the key.
+	type ipString string
+	allAddrs := map[ipString]*api.EndpointAddress{}
+	portsToAddrs := map[api.EndpointPort]addressSet{}
+	for i := range subsets {
+		for j := range subsets[i].Ports {
+			epp := &subsets[i].Ports[j]
+			for k := range subsets[i].Addresses {
```
Comment (bgrant0607) [inline pkg/api/endpoints/util.go]: Go question: Why not loop over the values instead of the keys?

Comment (satnam6502) [inline pkg/api/endpoints/util.go]: I think this is to avoid copying. The `range` construct has copy semantics.

Comment (thockin) [inline pkg/api/endpoints/util.go]: Looping over values causes copies.  Basically, any time there's a map or
slice where the value is anything but a primitive type you incur totally
pointless copies.  Go just encourages you to piss away your performance.

I asked them about this once and they simply said, "yeah, just iterate
slices in the C style, with a pointer to the current element".  So that's
the pattern I follow everywhere except test code where it doesn't matter.

In short `map[anything]struct` and `for x, y := range slice` are traps.

File: pkg/api/endpoints/util.go [hunk 15]
```diff
@@ -0,0 +1,154 @@
...
+	util.DeepHashObject(hasher, addrs)
+	return hex.EncodeToString(hasher.Sum(nil)[0:])
+}
+
+// SortSubsets sorts an array of EndpointSubset objects in place.  For ease of
+// use it returns the input slice.
+func SortSubsets(subsets []api.EndpointSubset) []api.EndpointSubset {
+	for i := range subsets {
+		ss := &subsets[i]
+		sort.Sort(addrsByHash(ss.Addresses))
```
Comment (bgrant0607) [inline pkg/api/endpoints/util.go]: Not that I care, but why by hash rather than by ip address?

Comment (thockin) [inline pkg/api/endpoints/util.go]: I just did the same thing for all.  I could easily change that to IP addr
if it bothers you even a little bit.

File: pkg/api/endpoints/util.go (hunk 14 above)
Comment (satnam6502) [inline pkg/api/endpoints/util.go]: Where is `for (const auto& kv` when you need it.

File: pkg/api/endpoints/util.go (hunk 15 above)
Comment (thockin) [inline pkg/api/endpoints/util.go]: pushed :)

File: pkg/api/endpoints/util.go (hunk 14 above)
Comment (thockin) [inline pkg/api/endpoints/util.go]: srsly, go need to offer iterate by ref.

File: pkg/api/endpoints/util.go [hunk 16]
```diff
@@ -0,0 +1,154 @@
...
+			for k := range subsets[i].Addresses {
+				epa := &subsets[i].Addresses[k]
+				ipstr := ipString(epa.IP)
+				// Accumulate the most "complete" address we can.
+				if allAddrs[ipstr] == nil {
+					// Make a copy so we don't write to the
+					// input args of this function.
+					p := &api.EndpointAddress{}
+					*p = *epa
+					allAddrs[ipstr] = p
```
Comment (bgrant0607) [inline pkg/api/endpoints/util.go]: Is this really better than DeepCopy?

File: pkg/api/endpoints/util_test.go [hunk 17]
```diff
@@ -0,0 +1,198 @@
...
+			}, {
+				Addresses: []api.EndpointAddress{{IP: "1.2.3.5"}},
+				Ports:     []api.EndpointPort{{Port: 333}},
+			}},
+			expect: []api.EndpointSubset{{
+				Addresses: []api.EndpointAddress{{IP: "1.2.3.4"}, {IP: "1.2.3.6"}},
+				Ports:     []api.EndpointPort{{Port: 111}},
+			}, {
+				Addresses: []api.EndpointAddress{{IP: "1.2.3.5"}},
+				Ports:     []api.EndpointPort{{Port: 222}, {Port: 333}},
```
Comment (bgrant0607) [inline pkg/api/endpoints/util_test.go]: I found this case surprising, but I suppose it's useful for roundtripping through the legacy representation.

File: pkg/api/endpoints/util.go (hunk 16 above)
Comment (thockin) [inline pkg/api/endpoints/util.go]: What is DeepCopy?  I find no such anything in our tree or in go's libs.
Sadly, this is just something go does not handle well.  No such thing as a
copy-ctor

File: pkg/api/v1beta1/conversion.go [hunk 18]
```diff
@@ -1279,32 +1300,10 @@ func init() {
...
-					if in.TargetRefs[j].Endpoint != in.Endpoints[i] {
-						continue
-					}
-					ep.TargetRef = &newer.ObjectReference{}
-					if err := s.Convert(&in.TargetRefs[j].ObjectReference, ep.TargetRef, 0); err != nil {
-						return err
-					}
-				}
-			}
+			// Back-compat fields are handled in the defaulting phase.
```
Comment (bgrant0607) [inline pkg/api/v1beta1/conversion.go]: Does the defaulting pass get invoked on the update path, or do you verify equality of the two representations?

File: pkg/api/v1beta1/defaults.go [hunk 19]
```diff
@@ -93,7 +96,42 @@ func init() {
 		},
 		func(obj *Endpoints) {
 			if obj.Protocol == "" {
-				obj.Protocol = "TCP"
+				obj.Protocol = ProtocolTCP
+			}
+			if len(obj.Subsets) == 0 && len(obj.Endpoints) > 0 {
```
Comment (bgrant0607) [inline pkg/api/v1beta1/defaults.go]: What should happen if both are nonempty?

File: pkg/api/v1beta1/conversion.go (hunk 18 above)
Comment (thockin) [inline pkg/api/v1beta1/conversion.go]: Defaulting runs any time we convert a versioned object into internal, which
should include updates.

File: pkg/api/v1beta1/defaults.go (hunk 19 above)
Comment (thockin) [inline pkg/api/v1beta1/defaults.go]: If they are both non-empty we should probably force the legacy fields to
match the subsets.  I think I just let some corner cases go here since they
are unlikely to happen (I think)

File: pkg/api/v1beta1/conversion.go [hunk 20]
```diff
@@ -1253,21 +1253,42 @@ func init() {
...
+					firstPortName = in.Subsets[0].Ports[0].Name
+				}
+			} else {
+				out.Protocol = ProtocolTCP
+			}
+			for i := range in.Subsets {
+				ss := &in.Subsets[i]
+				for j := range ss.Ports {
+					ssp := &ss.Ports[j]
+					if ssp.Name != firstPortName {
```
Comment (bgrant0607) [inline pkg/api/v1beta1/conversion.go]: I hope v1beta1 isn't around for too much longer, but is the expected behavior for v1beta1 documented somewhere?

* endpoints created via v1beta1 and converted to v1beta3
* endpoints created via v1beta3 and converted to v1beta1

File: pkg/registry/service/rest.go (hunk 13 above)
Comment (bgrant0607) [inline pkg/registry/service/rest.go]: To me, it only makes sense to permit the port to be omitted in the case that the service exposes just one port. Otherwise, we should require servicename:portnumber or servicename:portname.

**Comment by bgrant0607** [comment]
I'm done with my pass. LGTM for a first cut. Feel free to resolve feedback, as necessary, in subsequent PRs.

File: pkg/api/endpoints/util.go (hunk 16 above)
Comment (bgrant0607) [inline pkg/api/endpoints/util.go]: Sorry, nevermind. I thought I'd seen one, but it's not part of the standard library.

File: pkg/api/endpoints/util.go (hunk 14 above)
Comment (bgrant0607) [inline pkg/api/endpoints/util.go]: Ok, but I'll point out that the code a few lines below goes out of its way to copy every value.

**Comment by thockin** [comment]
Closes #5895

**Comment by satnam6502** [comment]
I will have one more careful look at the latest changes and if no further concerns I will merge by lunchtime.

File: pkg/api/v1beta1/conversion.go (hunk 20 above)
Comment (thockin) [inline pkg/api/v1beta1/conversion.go]: Added comments to v1b[12] about logic if one or the other type of endpoints are provided

File: pkg/registry/service/rest.go (hunk 13 above)
Comment (thockin) [inline pkg/registry/service/rest.go]: I don't think it is worth special casing one port - this allows you to have N ports, but only one can be unnamed.  OK?

File: pkg/api/endpoints/util.go (hunk 14 above)
Comment (thockin) [inline pkg/api/endpoints/util.go]: Go logic - the copy that is made on each loop has the same address, so I would need to make TWO copies in order to save one.

**Comment by thockin** [comment]
New push is up with all comments (AFAIK) addressed.  when this is OK, I need to squash this..

**Comment by satnam6502** [comment]
LGTM again and happy to merge after squash. After all, what can go wrong? It's Friday afternoon and e2e is on fire!

**Comment by satnam6502** [comment]
Is there a semi-plausible e2e run for the latest changes?

**Comment by thockin** [comment]
Squashed and pushed.  Running e2e now.

**Comment by thockin** [comment]
e2e passed except for elasticsearch, ahem.

**Comment by bgrant0607** [comment]
Merging.

File: pkg/service/endpoints_controller_test.go [hunk 21]
```diff
@@ -65,163 +62,156 @@ func newPodList(count int) *api.PodList {
...
+			Name:          "",
+			ContainerPort: 11,
+			Protocol:      "TCP",
+		}, {
+			Name:          "p",
+			ContainerPort: 22,
+			Protocol:      "TCP",
+		}}}},
+		port:     util.NewIntOrStringFromInt(93),
+		expected: 93,
```
Comment (hurf) [inline pkg/service/endpoints_controller_test.go]: Hi Tim, I have a question. In this case the container doesn't expose port 93. The service can still be created and the pod is bind to this service. But I tested it with following template:

service:

    protocol: TCP
    port: 999
    containerPort: 93

pod:
    
    ports:
          - name: testport
            protocol: TCP
            containerPort: 80
          - name: udp 
            protocol: UDP
            containerPort: 9999

I used `curl http://ip:999` to touch the service, it didn't work. But after I changed the serice with `containerPort: 80`, it worked.

Why we allow service connect to a port which the pod didn't expose? And there're more similar cases in this test.

Comment (thockin) [inline pkg/service/endpoints_controller_test.go]: Not every pod explicitly lists the ports it exposes.  It's a good idea to do so, for visibility, but it's not required (unless you want to give it a name)
